### PR TITLE
Implemented Append operation to append to existing files.

### PR DESCRIPTION
**Pull request by tyler-sommer**
This commit adds support for APPEND mode file operations on existing files
stored on HDFS.

### CODE REVIEW

File: file_writer.go [hunk 1]
```diff
@@ -46,6 +46,17 @@ func (c *Client) Create(name string) (*FileWriter, error) {
 	return c.CreateFile(name, replication, blockSize, 0644)
 }
 
+// Append opens an existing file in HDFS and returns an io.WriteCloser for
+// writing to it. Because of the way that HDFS writes are buffered and
+// acknowledged asynchronously, it is very important that Close is called after
+// all data has been written.
+//
+// Append is actually a proxy for AppendFile and exists for consistency with
+// Create and CreateFile.
```
Comment (colinmarc) [inline file_writer.go]: You can just have the one named `Append`, that's fine.

File: file_writer.go [hunk 2]
```diff
@@ -79,6 +90,54 @@ func (c *Client) CreateFile(name string, replication int, blockSize int64, perm
...
+
+	err = c.namenode.Execute("append", appendReq, appendResp)
+	if err != nil {
+		if nnErr, ok := err.(*rpc.NamenodeError); ok {
+			err = interpretException(nnErr.Exception, err)
+		}
+
+		return nil, &os.PathError{"append", name, err}
+	}
+	req := &hdfs.GetBlockLocationsRequestProto{
```
Comment (colinmarc) [inline file_writer.go]: Probably add a newline before this line, since it's a new action.

**Comment by colinmarc** [comment]
Awesome - thanks for the contribution! This looks great, but can you add some tests?

File: client_test.go [hunk 3]
```diff
@@ -101,6 +101,7 @@ func TestCopyToLocal(t *testing.T) {
 func TestCopyToRemote(t *testing.T) {
 	client := getClient(t)
 
+	baleet(t, "/_test/copytoremote.txt")
```
Comment (tyler-sommer) [inline client_test.go]: This was added to prevent errors on multiple test runs against the same hdfs cluster.

**Comment by tyler-sommer** [comment]
I made the requested changes, but the tests are failing on hdp with

```

--- FAIL: TestFileAppend-2 (0.03s)
	Error Trace:	file_writer_test.go:214
	Error:		Received unexpected error "create /_test/append/1.txt: addBlock call failed with ERROR_APPLICATION (org.apache.hadoop.hdfs.server.namenode.NotReplicatedYetException)"
		
=== RUN TestFileAppendEmptyFile-2
--- FAIL: TestFileAppendEmptyFile-2 (0.03s)
	Error Trace:	file_writer_test.go:243
	Error:		Received unexpected error "create /_test/append/1.txt: addBlock call failed with ERROR_APPLICATION (org.apache.hadoop.hdfs.server.namenode.NotReplicatedYetException)"
```

Would adding a `time.Sleep()` call be acceptable?

File: file_writer_test.go [hunk 4]
```diff
@@ -192,3 +192,68 @@ func TestCreateEmptyFileWithoutPermission(t *testing.T) {
...
+	err = writer.Close()
+	require.NoError(t, err)
+
+	reader, err := client.Open("/_test/append/1.txt")
+	require.NoError(t, err)
+
+	bytes, err := ioutil.ReadAll(reader)
+	require.NoError(t, err)
+	assert.Equal(t, "foobar", string(bytes))
+}
```
Comment (colinmarc) [inline file_writer_test.go]: Can you add a trailing newline here? sorry to nitpick

**Comment by colinmarc** [comment]
Looks like the error is specific to HDP, which is fun. Maybe try bumping the version of the hadoop tarball we use?

**Comment by tyler-sommer** [comment]
Hmm, updating to HDP 2.2's hadoop 2.6 tarball resolved the issue on my OSX machine. I'll dig a bit more to try and get this worked out.

Sorry about all the notifications!

**Comment by tyler-sommer** [comment]
Updating to HDP 2.4's hadoop 2.7 did the trick. I do wonder if this points to some issue in specific configurations. 

Thanks for your time, I really appreciate it!

**Comment by colinmarc** [comment]
Thanks so much for the contribution! Can you rebase one more time? I made the go version change as a separate commit just now.

**Comment by tyler-sommer** [comment]
Done deal! Thanks for this great project, it's already proved very useful. Cheers

**Comment by deepix** [comment]
I'm getting the same exception (`NotReplicatedYetException`).  I'm on stock Hadoop 2.6.3.  Would you recommend an upgrade to 2.7?  Thanks!

**Pull request merged**